### PR TITLE
Create Item repository and controller

### PR DESCRIPTION
**Pull request by lhenz15**

### CODE REVIEW

File: src/main/java/repositories/ItemRepository.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
+package repositories;
+
+import entities.Item;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class ItemRepository implements IGenericRepository<Item> {
+
+    Map<Integer, Item> items;
```
Comment (reynoldmorel) [inline src/main/java/repositories/ItemRepository.java]: Could make this private?

Comment (reynoldmorel) [reply src/main/java/repositories/ItemRepository.java]: Also I think you can initialize it directly

File: src/main/java/repositories/ItemRepository.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+    public ItemRepository() {
+        items = new HashMap<>();
+    }
+
+    @Override
+    public void add(Item item) {
+        items.put(item.getId(), item);
```
Comment (reynoldmorel) [inline src/main/java/repositories/ItemRepository.java:20]: You should generate the id within the repository

File: src/main/java/repositories/ItemRepository.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    @Override
+    public Item update(Item item) {
+        if (item != null && items.containsKey(item.getId())) {
+            return items.replace(item.getId(), item);
```
Comment (reynoldmorel) [inline src/main/java/repositories/ItemRepository.java]: You should throw some sort of item not found exception if item is not found. Also you should return the updated item

File: src/main/java/repositories/ItemRepository.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        }
+
+        return null;
+    }
+
+    @Override
+    public Item delete(Integer id) {
+        if (id != null && items.containsKey(id)) {
+            return items.remove(id);
```
Comment (reynoldmorel) [inline src/main/java/repositories/ItemRepository.java]: Same here if item is not found just throw an exception

File: src/main/java/repositories/ItemRepository.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        }
+
+        return null;
+    }
+
+    @Override
+    public Item find(Integer id) {
+        if (id != null && items.containsKey(id)) {
+            return items.get(id);
```
Comment (reynoldmorel) [inline src/main/java/repositories/ItemRepository.java]: Use getOrDefsult()

File: src/main/java/repositories/ItemRepository.java [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+        if (id != null && items.containsKey(id)) {
+            return items.remove(id);
+        }
+
+        return null;
+    }
+
+    @Override
+    public Item find(Integer id) {
+        if (id != null && items.containsKey(id)) {
```
Comment (reynoldmorel) [inline src/main/java/repositories/ItemRepository.java]: You don't need this check

File: src/main/java/repositories/ItemRepository.java (hunk 4 above)
Comment (reynoldmorel) [inline src/main/java/repositories/ItemRepository.java]: You don't need this check

File: src/main/java/repositories/ItemRepository.java [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+            return items.get(id);
+        }
+
+        return null;
+    }
+
+    @Override
+    public Iterable<Item> findAll() {
+        return items.values();
```
Comment (reynoldmorel) [inline src/main/java/repositories/ItemRepository.java]: Return a list from here. Just do stream() and collect()

**Review by reynoldmorel — CHANGES_REQUESTED** [review]

File: src/main/java/repositories/ItemRepository.java (hunk 5 above)
Comment (lhenz15) [reply src/main/java/repositories/ItemRepository.java]: What should I return by default? an empty Item?

File: src/main/java/controllers/ItemController.java [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+import entities.Item;
+import repositories.ItemRepository;
+
+import java.util.List;
+
+public class ItemController {
+
+    private ItemRepository itemRepository;
+
+    public ItemController() {
```
Comment (reynoldmorel) [inline src/main/java/controllers/ItemController.java]: Inject `itemRepository` throught the constructor.

File: src/main/java/repositories/IGenericRepository.java [hunk 9]
```diff
@@ -0,0 +1,13 @@
+package repositories;
+
+public interface IGenericRepository<T> {
+    void add(T t);
+
+    T update(T t);
+
+    T delete(Integer id);
```
Comment (reynoldmorel) [inline src/main/java/repositories/IGenericRepository.java]: What if I want the key to be `String`?

File: src/main/java/repositories/ItemRepository.java [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+import utils.NotFoundException;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class ItemRepository implements IGenericRepository<Item> {
+
+    private Map<Integer, Item> items = new HashMap<>();
```
Comment (reynoldmorel) [inline src/main/java/repositories/ItemRepository.java]: Shouldn't this be final?

File: src/main/java/repositories/ItemRepository.java [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+    private int itemId = 1;
+
+    @Override
+    public void add(Item item) {
+        item.setId(itemId);
```
Comment (reynoldmorel) [inline src/main/java/repositories/ItemRepository.java]: I think you can do `item++` directly here

File: src/main/java/repositories/ItemRepository.java [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+
+        itemId++;
+    }
+
+    @Override
+    public Item update(Item item) {
+        if(items.replace(item.getId(), item) != null)
+            return item;
+
+        throw new NotFoundException("Item not found.");
```
Comment (reynoldmorel) [inline src/main/java/repositories/ItemRepository.java]: Let's rename this to `EntityNotFoundException`

File: src/main/java/repositories/ItemRepository.java [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+
+import entities.Item;
+import utils.NotFoundException;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class ItemRepository implements IGenericRepository<Item> {
```
Comment (reynoldmorel) [inline src/main/java/repositories/ItemRepository.java]: nit:
```
return Optional.ofNullable(items.get(id)).orElseThrow(EntityNotFoundException::new)
```

File: src/main/java/utils/NotFoundException.java [hunk 14]
```diff
@@ -0,0 +1,7 @@
+package utils;
+
+public class NotFoundException extends RuntimeException {
+    public NotFoundException(String message) {
+        super(message);
```
Comment (reynoldmorel) [inline src/main/java/utils/NotFoundException.java]: Use default message.

**Review by reynoldmorel — APPROVED** [review]
Nit: could you change generic T to K since looks like K represents better the key data type

**Pull request merged**